examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs

```diff
@@ -0,0 +1,191 @@
+use core::marker::PhantomData;
+use ink_primitives::{
+    Key,
+    KeyPtr,
+};
+use ink_storage::traits::{
+    PackedAllocate,
+    PackedLayout,
+    SpreadAllocate,
+    SpreadLayout,
+};
+use scale::{
+    Decode,
+    Encode,
+};
+
+/// It is a status struct for `Upgradeable`, to specify that the inner type is initialized.
+#[derive(Debug)]
+pub struct Initialized;
+/// It is a status struct for `Upgradeable`, to specify that the inner type may be not initialized and
+/// `pull_spread` should initialize it.
+#[derive(Debug)]
+pub struct NotInitialized;
+
+/// The `Upgradeable` means if the field is not initialized, it will be.
+///
+/// By default ink! would throw an error that the field is not initialized.
+/// With that wrapper, you can initialize the field later during the method execution,
+/// not in the constructor. It can be done because `SpreadLayout` for `Upgradeable<T, NotInitialized>`
+/// creates the object, if storage key is empty.
+#[derive(Debug, Decode, Encode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct Upgradeable<T: PackedLayout, InitializationStatus = Initialized> {
+    inner: T,
+    status: PhantomData<fn() -> InitializationStatus>,
+}
+
+impl<T: PackedLayout, State> Upgradeable<T, State> {
+    pub fn new(inner: T) -> Self {
+        Upgradeable {
+            inner,
+            status: Default::default(),
+        }
+    }
+}
+
+/// It is default implementation of `SpreadLayout` for case when we don't need to init.
+impl<T: PackedLayout> SpreadLayout for Upgradeable<T, Initialized> {
+    const FOOTPRINT: u64 = T::FOOTPRINT;
+    const REQUIRES_DEEP_CLEAN_UP: bool = T::REQUIRES_DEEP_CLEAN_UP;
+
+    fn pull_spread(ptr: &mut KeyPtr) -> Self {
+        Upgradeable::new(T::pull_spread(ptr))
+    }
+
+    fn push_spread(&self, ptr: &mut KeyPtr) {
+        T::push_spread(&self.inner, ptr)
+    }
+
+    fn clear_spread(&self, ptr: &mut KeyPtr) {
+        T::clear_spread(&self.inner, ptr)
+    }
+}
+
+/// It is implementation of `SpreadLayout` that initialize the inner type if it is not initialized.
+impl<T: PackedLayout + SpreadAllocate> SpreadLayout for Upgradeable<T, NotInitialized> {
+    const FOOTPRINT: u64 = <T as SpreadLayout>::FOOTPRINT;
+    const REQUIRES_DEEP_CLEAN_UP: bool = <T as SpreadLayout>::REQUIRES_DEEP_CLEAN_UP;
+
+    fn pull_spread(ptr: &mut KeyPtr) -> Self {
+        if ink_env::get_contract_storage::<T>(ptr.advance_by(0))
```

Comment:
```suggestion
        if ink_env::get_contract_storage::<T>(ptr.key())
```

Reply:
`key` returns the previous key, we need here an actual key, so we are forcing `advance_by(0)`.
<img width="811" alt="image" src="https://user-images.githubusercontent.com/18346821/156986730-83dc58e2-dc8a-42fa-aa05-b97148f592eb.png">